primitives/primproc/batchprimitiveprocessor.cpp

```diff
@@ -1384,8 +1392,18 @@ void BatchPrimitiveProcessor::executeTupleJoin()
             //else
             // cout << "j != joinerCount\n";
         }
+        // If we've accumulated more than 1048576 (2^20)_ of resultCounts, cut off processing. The caller will
+        // restart to continue where we left off.
+        if (resultCount >= 1048576)
```

Comment:
Why did we add this condition and what is the rationale behind the literal?

Reply:
This can get complicated.
Each time Execute() is called, it passed a bloc of up to 8192 rows of a table. Somewhere in there executeTupleJoin() is called. It looks for all the join matches for each row and builds a data set in tSmallSideMatches. This little multi-dimensional array has a first dimension as the number of joins in the query (joinerCount) and the second dimension is a vector of match results for each row up 8192 for the first join and up to sum of matchCount for each of those 8192 for the second join, compounded for the third, etc. Each of these have unlimited number of entries for the final join.. In the test case, in the hundreds of thousands. The result can get extremely large.

After tSmallSideMatches is built, it either sends it back to ExeMgr along with the raw data, or if there's funcexp or aggregation, materializes the rows and performs the operations and sends it ExeMgr. In the later case, there was logic to limit package size, but in the former, there was no limit to package size. I was seeing packets in range of 2-3GB.

So, my changes limit the size of tSmallSideMatches, and thus the size of everything else being done. resultCount reflects the size of tSmallSideMatches. This limiting number has to be smaller than OOM and larger than 8192. I chose 2^10 as a good compromise. The condition says we have enough data. Return what we have so far.

The code in Execute() has been modified to handle this smaller count and loops to call executeTupleJoin() again with the new starting position until all 8192 initial rows are processed.